# Profitable App Profiles for the App Store and Google Play Markets
Our aim in this project is to find mobile app profiles that are profitable for the App Store and Google Play markets. We're working as data analysts for a company that builds Android and iOS mobile apps, and our job is to enable our team of developers to make data-driven decisions with respect to the kind of apps they build.

At our company, we only build apps that are free to download and install, and our main source of revenue consists of in-app ads. This means that our revenue for any given app is mostly influenced by the number of users that use our app. Our goal for this project is to analyze data to help our developers understand what kinds of apps are likely to attract more users.
## Opening and Exploring the Data
As of September 2018, there were approximately 2 million iOS apps available on the App Store, and 2.1 million Android apps on Google Play.

Collecting data for over four million apps requires a significant amount of time and money, so we'll try to analyze a sample of data instead. To avoid spending resources with collecting new data ourselves, we should first try to see whether we can find any relevant existing data at no cost. Luckily, these are two data sets that seem suitable for our purpose:

- [A data set](https://www.kaggle.com/lava18/google-play-store-apps) containing data about approximately ten thousand Android apps from Google Play. You can download the data set directly from [this link](https://dq-content.s3.amazonaws.com/350/googleplaystore.csv);
- [A data set](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps) containing data about approximately seven thousand iOS apps from the App Store. You can download the data set directly from [this link](https://dq-content.s3.amazonaws.com/350/AppleStore.csv).

Let's start by opening the two data sets and then continue with exploring the data.

In [1]:
from csv import reader
# The Google Play data set
abrir_ficheiro = open('googleplaystore.csv', encoding="utf8")
ler_ficheiro = reader(abrir_ficheiro)
android = list(ler_ficheiro)
android_header = android[0]
android = android[1:]

# The App Store data set
abrir_ficheiro = open('Applestore.csv', encoding="utf8")
ler_ficheiro = reader(abrir_ficheiro)
ios = list(ler_ficheiro)
ios_header = ios[0]
ios = ios[1:]

To make it easier to explore the two data sets, we'll first write a function named explore_data() that we can use repeatedly to explore rows in a more readable way. We'll also add an option for our function to show the number of rows and columns for any data set.

In [2]:
def explore_data(dados, começo, fim, linhas_colunas=False):
    for linha in dados[começo:fim]:
        print(linha)
        print('\n')
    if linhas_colunas:
        print(f'\033[1;30mNúmero de linhas = {len(dados)}')
        print(f'Número de colunas = {len(dados[0])} \033[m')


print('\033[1;34m', android_header, '\033[m')
print('\n')
análise_android = explore_data(android, 0, 3, linhas_colunas=True)

 ['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'] 


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Número de linhas = 10841
Número de colunas = 13 


We see that the Google Play data set has 10841 apps and 13 columns. At a quick glance, the columns that might be useful for the purpose of our analysis are 'App', 'Category', 'Reviews', 'Installs', 'Type', 'Price', and 'Genres'.

Now let's take a look at the App Store data set.

In [3]:
print('\033[1;34m', ios_header, '\033[m')
print('\n')
análise_ios = explore_data(ios, 0, 3, linhas_colunas=True)

 ['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'] 


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Número de linhas = 7197
Número de colunas = 16 


We have 7197 iOS apps in this data set, and the columns that seem interesting are: 'track_name', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', and 'prime_genre'. Not all column names are self-explanatory in this case, but details about each column can be found in the data set [documentation](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps).

## Deleting Wrong Data
The Google Play data set has a dedicated [discussion section](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/discussion), and we can see that [one of the discussions](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015) outlines an error for row 10472. Let's print this row and compare it against the header and another row that is correct.

In [4]:
print(android[10472]) # Incorrect row
print('\n')
print(android_header) # Header
print('\n')
print(android[0]) # Example of one correct row

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


The row 10472 corresponds to the app Life Made WI-Fi Touchscreen Photo Frame, and we can see that the rating is 19. This is clearly off because the maximum rating for a Google Play app is 5 (as mentioned in the [discussions section](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015), this problem is caused by a missing value in the 'Category' column). As a consequence, we'll delete this row.

Alternatively, we could easily search the app category and use the insert function to fix this data point.

In [5]:
print(len(android))
android.pop(10472)
print(len(android))

10841
10840


## Removing Duplicate Entries
### Part one
If we explore the Google Play data set long enough, we'll find that some apps have more than one entry. For instance, the application Instagram has four entries:

In [6]:
for app in android:
    if app[0] == 'Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


In total, there are 1,181 cases where an app occurs more than once:

In [7]:
únicas = []
repetidas = []

for app in android:
    if app[0] not in únicas:
        únicas.append(app[0])
    else:
        repetidas.append(app[0])

print(f'Número de aplicações repetidas: {len(repetidas)} ')
print('\n')
print('Exemplos de aplicações repetidas:')
for app in repetidas[:5]:
    print('\n')
    print('    ', app)
print('\n')
print('               ...')

Número de aplicações repetidas: 1181 


Exemplos de aplicações repetidas:


     Quick PDF Scanner + OCR FREE


     Box


     Google My Business


     ZOOM Cloud Meetings


     join.me - Simple Meetings


               ...


We don't want to count certain apps more than once when we analyze data, so we need to remove the duplicate entries and keep only one entry per app. One thing we could do is remove the duplicate rows randomly, but we could probably find a better way.

If you examine the rows we printed two cells above for the Instagram app, the main difference happens on the fourth position of each row, which corresponds to the number of reviews. The different numbers show that the data was collected at different times. We can use this to build a criterion for keeping rows. We won't remove rows randomly, but rather we'll keep the rows that have the highest number of reviews because the higher the number of reviews, the more reliable the ratings.

To do that, we will:

- Create a dictionary where each key is a unique app name, and the value is the highest number of reviews of that app;
- Use the dictionary to create a new data set, which will have only one entry per app (and we only select the apps with the highest number of reviews).

### Part Two
Let's start by building the dictionary.

In [8]:
reviews_max = {}

for app in android:
    name = app[0]
    n_reviews = float(app[3])
    
    if name not in reviews_max:
        reviews_max[name] = n_reviews
    else:
        if reviews_max[name] < n_reviews:
            reviews_max[name] = n_reviews

In a previous code cell, we found that there are 1,181 cases where an app occurs more than once, so the length of our dictionary (of unique apps) should be equal to the difference between the length of our data set and 1,181.

In [9]:
print(f'Expected lenght: {len(android) - 1181}')
print(f'Actual lenght: {len(reviews_max)}')

Expected lenght: 9659
Actual lenght: 9659


Now, let's use the reviews_max dictionary to remove the duplicates. For the duplicate cases, we'll only keep the entries with the highest number of reviews. In the code cell below:

- We start by initializing two empty lists, android_clean and already_added;
- We loop through the android data set, and for every iteration:
 -   We isolate the name of the app and the number of reviews.
 -   We add the current row (app) to the android_clean list, and the app name (name) to the already_added list if:
     - The number of reviews of the current app matches the number of reviews of that app as described in the reviews_max dictionary; and
     - The name of the app is not already in the already_added list. We need to add this supplementary condition to account for those cases where the highest number of reviews of a duplicate app is the same for more than one entry (for example, the Box app has three entries, and the number of reviews is the same). If we just check for reviews_max[name] == n_reviews, we'll still end up with duplicate entries for some apps.

In [10]:
android_clean = []
already_added = []

for app in android:
    name = app[0]
    n_reviews = float(app[3])
    
    if reviews_max[name] == n_reviews and name not in already_added:
        android_clean.append(app)
        already_added.append(name)

Now let's quickly explore the new data set, and confirm that the number of rows is 9,659.

In [11]:
explore_data(android_clean, 0, 3, linhas_colunas=True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Número de linhas = 9659
Número de colunas = 13 


We have 9659 rows, just as expected.

In the previous step, we managed to remove the duplicate app entries in the Google Play data set. We don't need to do the same for the App Store data because there are no duplicates — we can easily check that by using the id column (which is available on this data set, as opposed to what we have seen for Google Play data set.

## Removing Non-English Apps
### Part One

If you explore the data sets enough, you'll notice the names of some of the apps suggest they are not directed toward an English-speaking audience. Below, we see a couple of examples from both data sets:

In [12]:
print(ios[813][1])
print(ios[6731][1])

print(android_clean[4412][0])
print(android_clean[7940][0])

爱奇艺PPS -《欢乐颂2》电视剧热播
【脱出ゲーム】絶対に最後までプレイしないで 〜謎解き＆ブロックパズル〜
中国語 AQリスニング
لعبة تقدر تربح DZ


We're not interested in keeping these kind of apps, so we'll remove them. One way to go about this is to remove each app whose name contains a symbol that is not commonly used in English text — English text usually includes letters from the English alphabet, numbers composed of digits from 0 to 9, punctuation marks (., !, ?, ;, etc.), and other symbols (+, *, /, etc.).

All these characters that are specific to English texts are encoded using the ASCII standard. Each ASCII character has a corresponding number between 0 and 127 associated with it, and we can take advantage of that to build a function that checks an app name and tells us whether it contains non-ASCII characters.

We built this function below, and we use the built-in ord() function to find out the corresponding encoding number of each character.

In [13]:
def english_filter(txt):
    
    for character in txt:
        if ord(character) > 127:
            return False
    
    return True

teste1 = english_filter('Instagram')
teste2 = english_filter('爱奇艺PPS -《欢乐颂2》电视剧热播')
print(teste1)
print(teste2)

True
False


The function seems to work fine, but some English app names use emojis or other symbols (™, — (em dash), – (en dash), etc.) that fall outside of the ASCII range. Because of this, we'll remove useful apps if we use the function in its current form, as we can see below:

In [14]:
print(english_filter('Docs To Go™ Free Office Suite'))
print(english_filter('Instachat 😜'))

print(ord('™'))
print(ord('😜'))

False
False
8482
128540


### Part Two

To minimize the impact of data loss, we'll only remove an app if its name has more than three non-ASCII characters. For that reason, let's update the function writen above:

In [15]:
def english_filter(txt):
    
    contador = 0
    for character in txt:
        if ord(character) > 127:
            contador += 1
    
    if contador > 3:
        return False
    else:
        return True


print(english_filter('Docs To Go™ Free Office Suite'))
print(english_filter('Instachat 😜'))

True
True


The function is still not perfect, and very few non-English apps might get past our filter, but this seems good enough at this point in our analysis — we shouldn't spend too much time on optimization at this point.

Below, we use the english_filter() function to filter out the non-English apps for both data sets:

In [16]:
android_english = []
ios_english = []

for app in android_clean:
    name = app[0]
    if english_filter(name):
        android_english.append(app)

for app in ios:
    name = app[1]
    if english_filter(name):
        ios_english.append(app)

explore_data(android_english, 0, 3, True)
print('\n')
explore_data(ios_english, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Número de linhas = 9614
Número de colunas = 13 


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '

We can see that we're left with 9614 Android apps and 6183 iOS apps.

## Isolating the Free Apps
As we mentioned in the introduction, we only build apps that are free to download and install, and our main source of revenue consists of in-app ads. Our data sets contain both free and non-free apps, and we'll need to isolate only the free apps for our analysis. Below, we isolate the free apps for both our data sets.

In [17]:
android_final = []
ios_final = []

for app in android_english:
    price = app[7]
    if price == '0':
        android_final.append(app)

for app in ios_english:
    price = float(app[4])
    if price == 0:
        ios_final.append(app)

print(len(android_final))
print(len(ios_final))

8864
3222


We're left with 8864 Android apps and 3222 iOS apps, which should be enough for our analysis.

## Most Common Apps by Genre
### Part One
As we mentioned in the introduction, our aim is to determine the kinds of apps that are likely to attract more users because our revenue is highly influenced by the number of people using our apps.

To minimize risks and overhead, our validation strategy for an app idea is comprised of three steps:

1. Build a minimal Android version of the app, and add it to Google Play;
2. If the app has a good response from users, we then develop it further;
3. If the app is profitable after six months, we also build an iOS version of the app and add it to the App Store.

Because our end goal is to add the app on both the App Store and Google Play, we need to find app profiles that are successful on both markets. For instance, a profile that might work well for both markets might be a productivity app that makes use of gamification.

Let's begin the analysis by getting a sense of the most common genres for each market. For this, we'll build a frequency table for the prime_genre column of the App Store data set, and the Genres and Category columns of the Google Play data set.

### Part Two
Now we will build one function that we can use to analyze the frequency tables. This function has two purposes:
1. Generate frequency tables that show percentages;
2. Display the percentages in a descending order.

In [18]:
def freq_table(dataset, index):
    from operator import itemgetter
    
    f_table = {}
    total = 0
    for app in dataset:
        entrada = app[index]
        total += 1
        
        if entrada not in f_table:
            f_table[entrada] = 1
        else:
            f_table[entrada] += 1
    
    for k, v in f_table.items():
        f_table[k] = (v / total) * 100
    
    table_sorted = sorted(f_table.items(), key=itemgetter(1), reverse=True)
    # Organiza o dicionário f_table de modo decrescente, de acordo com a percentagem.
    
    for app in table_sorted:
        print(f"{app[0]} = {app[1]:.2f}%")

### Part Three
We start by examining the frequency table for the prime_genre column of the App Store data set.

In [19]:
freq_table(ios_final, -5)

Games = 58.16%
Entertainment = 7.88%
Photo & Video = 4.97%
Education = 3.66%
Social Networking = 3.29%
Shopping = 2.61%
Utilities = 2.51%
Sports = 2.14%
Music = 2.05%
Health & Fitness = 2.02%
Productivity = 1.74%
Lifestyle = 1.58%
News = 1.33%
Travel = 1.24%
Finance = 1.12%
Weather = 0.87%
Food & Drink = 0.81%
Reference = 0.56%
Business = 0.53%
Book = 0.43%
Navigation = 0.19%
Medical = 0.19%
Catalogs = 0.12%


We can see that among the free English apps, more than a half (58.16%) are games. Entertainment apps are close to 8%, followed by photo and video apps, which are close to 5%. Only 3.66% of the apps are designed for education, followed by social networking apps which amount for 3.29% of the apps in our data set.

The general impression is that App Store (at least the part containing free English apps) is dominated by apps that are designed for fun (games, entertainment, photo and video, social networking, sports, music, etc.), while apps with practical purposes (education, shopping, utilities, productivity, lifestyle, etc.) are more rare. However, the fact that fun apps are the most numerous doesn't also imply that they also have the greatest number of users — the demand might not be the same as the offer.

Let's continue by examining the Category and Genres columns of the Google Play data set (two columns which seem to be related).

In [20]:
freq_table(android_final, 1) # Category

FAMILY = 18.91%
GAME = 9.72%
TOOLS = 8.46%
BUSINESS = 4.59%
LIFESTYLE = 3.90%
PRODUCTIVITY = 3.89%
FINANCE = 3.70%
MEDICAL = 3.53%
SPORTS = 3.40%
PERSONALIZATION = 3.32%
COMMUNICATION = 3.24%
HEALTH_AND_FITNESS = 3.08%
PHOTOGRAPHY = 2.94%
NEWS_AND_MAGAZINES = 2.80%
SOCIAL = 2.66%
TRAVEL_AND_LOCAL = 2.34%
SHOPPING = 2.25%
BOOKS_AND_REFERENCE = 2.14%
DATING = 1.86%
VIDEO_PLAYERS = 1.79%
MAPS_AND_NAVIGATION = 1.40%
FOOD_AND_DRINK = 1.24%
EDUCATION = 1.16%
ENTERTAINMENT = 0.96%
LIBRARIES_AND_DEMO = 0.94%
AUTO_AND_VEHICLES = 0.93%
HOUSE_AND_HOME = 0.82%
WEATHER = 0.80%
EVENTS = 0.71%
PARENTING = 0.65%
ART_AND_DESIGN = 0.64%
COMICS = 0.62%
BEAUTY = 0.60%


The landscape seems significantly different on Google Play: there are not that many apps designed for fun, and it seems that a good number of apps are designed for practical purposes (family, tools, business, lifestyle, productivity, etc.). However, if we investigate this further, we can see that the family category (which accounts for almost 19% of the apps) means mostly games for kids.

<img src="android_family apps.png">

Even so, practical apps seem to have a better representation on Google Play compared to App Store. This picture is also confirmed by the frequency table we see for the Genres column:

In [21]:
freq_table(android_final, -4)

Tools = 8.45%
Entertainment = 6.07%
Education = 5.35%
Business = 4.59%
Lifestyle = 3.89%
Productivity = 3.89%
Finance = 3.70%
Medical = 3.53%
Sports = 3.46%
Personalization = 3.32%
Communication = 3.24%
Action = 3.10%
Health & Fitness = 3.08%
Photography = 2.94%
News & Magazines = 2.80%
Social = 2.66%
Travel & Local = 2.32%
Shopping = 2.25%
Books & Reference = 2.14%
Simulation = 2.04%
Dating = 1.86%
Arcade = 1.85%
Video Players & Editors = 1.77%
Casual = 1.76%
Maps & Navigation = 1.40%
Food & Drink = 1.24%
Puzzle = 1.13%
Racing = 0.99%
Libraries & Demo = 0.94%
Role Playing = 0.94%
Auto & Vehicles = 0.93%
Strategy = 0.91%
House & Home = 0.82%
Weather = 0.80%
Events = 0.71%
Adventure = 0.68%
Comics = 0.61%
Art & Design = 0.60%
Beauty = 0.60%
Parenting = 0.50%
Card = 0.45%
Casino = 0.43%
Trivia = 0.42%
Educational;Education = 0.39%
Board = 0.38%
Educational = 0.37%
Education;Education = 0.34%
Word = 0.26%
Casual;Pretend Play = 0.24%
Music = 0.20%
Entertainment;Music & Video = 0.17%
Puzzle

The difference between the Genres and the Category columns is not crystal clear, but one thing we can notice is that the Genres column is much more granular (it has more categories). We're only looking for the bigger picture at the moment, so we'll only work with the Category column moving forward.

Up to this point, we found that the App Store is dominated by apps designed for fun, while Google Play shows a more balanced landscape of both practical and for-fun apps. Now we'd like to get an idea about the kind of apps that have most users.

## Most Popular Apps by Genre on the App Store

One way to find out what genres are the most popular (have the most users) is to calculate the average number of installs for each app genre. For the Google Play data set, we can find this information in the Installs column, but for the App Store data set this information is missing. As a workaround, we'll take the total number of user ratings as a proxy, which we can find in the rating_count_tot app.

Below, we calculate the average number of user ratings per app genre on the App Store:

In [22]:
def f_table(dataset, index):
    fr_table = {}
    total = 0
    for app in dataset:
        entrada = app[index]
        total += 1
        
        if entrada not in fr_table:
            fr_table[entrada] = 1
        else:
            fr_table[entrada] += 1
    
    for k, v in fr_table.items():
        fr_table[k] = (v / total) * 100
    
    return fr_table

In [23]:
genre_ios = f_table(ios_final, -5)

for genre in genre_ios:
    total = 0
    len_genre = 0
    for app in ios_final:
        genre_app = app[-5]
        
        if genre_app == genre:
            rating = float(app[5])
            total += rating
            len_genre += 1
    print(f"Avg_{genre} = {total / len_genre:.2f}")


Avg_Social Networking = 71548.35
Avg_Photo & Video = 28441.54
Avg_Games = 22788.67
Avg_Music = 57326.53
Avg_Reference = 74942.11
Avg_Health & Fitness = 23298.02
Avg_Weather = 52279.89
Avg_Utilities = 18684.46
Avg_Travel = 28243.80
Avg_Shopping = 26919.69
Avg_News = 21248.02
Avg_Navigation = 86090.33
Avg_Lifestyle = 16485.76
Avg_Entertainment = 14029.83
Avg_Food & Drink = 33333.92
Avg_Sports = 23008.90
Avg_Book = 39758.50
Avg_Finance = 31467.94
Avg_Education = 7003.98
Avg_Productivity = 21028.41
Avg_Business = 7491.12
Avg_Catalogs = 4004.00
Avg_Medical = 612.00


On average, navigation apps have the highest number of user reviews, but this figure is heavily influenced by Waze and Google Maps, which have close to half a million user reviews together:

In [24]:
for app in ios_final:
    genre = app[-5]
    if genre == 'Navigation':
        print(f"{app[1]} : {app[5]}")
        

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5


The same pattern applies to social networking apps, where the average number is heavily influenced by a few giants like Facebook, Pinterest, Skype, etc. Same applies to music apps, where a few big players like Pandora, Spotify, and Shazam heavily influence the average number.

Our aim is to find popular genres, but navigation, social networking or music apps might seem more popular than they really are. The average number of ratings seem to be skewed by very few apps which have hundreds of thousands of user ratings, while the other apps may struggle to get past the 10,000 threshold. We could get a better picture by removing these extremely popular apps for each genre and then rework the averages, but we'll leave this level of detail for later.

Reference apps have 74,942 user ratings on average, but it's actually the Bible and Dictionary.com which skew up the average rating:

In [25]:
for app in ios_final:
    genre = app[-5]
    if genre == 'Reference':
        print(f"{app[1]} : {app[5]}")
        

Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
教えて!goo : 0
Jishokun-Japanese English Dictionary & Translator : 0


However, this niche seems to show some potential. One thing we could do is take another popular book and turn it into an app where we could add different features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes about the book, etc. On top of that, we could also embed a dictionary within the app, so users don't need to exit our app to look up words in an external app.

This idea seems to fit well with the fact that the App Store is dominated by for-fun apps. This suggests the market might be a bit saturated with for-fun apps, which means a practical app might have more of a chance to stand out among the huge number of apps on the App Store.

Other genres that seem popular include weather, book, food and drink, or finance. The book genre seem to overlap a bit with the app idea we described above, but the other genres don't seem too interesting to us:

Weather apps — people generally don't spend too much time in-app, and the chances of making profit from in-app adds are low. Also, getting reliable live weather data may require us to connect our apps to non-free APIs.

Food and drink — examples here include Starbucks, Dunkin' Donuts, McDonald's, etc. So making a popular food and drink app requires actual cooking and a delivery service, which is outside the scope of our company.

Finance apps — these apps involve banking, paying bills, money transfer, etc. Building a finance app requires domain knowledge, and we don't want to hire a finance expert just to build an app.

Now let's analyze the Google Play market a bit.

## Most Popular Apps by Genre on Google Play¶
For the Google Play market, we actually have data about the number of installs, so we should be able to get a clearer picture about genre popularity. However, the install numbers don't seem precise enough — we can see that most values are open-ended (100+, 1,000+, 5,000+, etc.):

In [26]:
freq_table(android_final, 5) # the Install columns

1,000,000+ = 15.73%
100,000+ = 11.55%
10,000,000+ = 10.55%
10,000+ = 10.20%
1,000+ = 8.39%
100+ = 6.92%
5,000,000+ = 6.83%
500,000+ = 5.56%
50,000+ = 4.77%
5,000+ = 4.51%
10+ = 3.54%
500+ = 3.25%
50,000,000+ = 2.30%
100,000,000+ = 2.13%
50+ = 1.92%
5+ = 0.79%
1+ = 0.51%
500,000,000+ = 0.27%
1,000,000,000+ = 0.23%
0+ = 0.05%
0 = 0.01%


One problem with this data is that is not precise. For instance, we don't know whether an app with 100,000+ installs has 100,000 installs, 200,000, or 350,000. However, we don't need very precise data for our purposes — we only want to get an idea which app genres attract the most users, and we don't need perfect precision with respect to the number of users.

We're going to leave the numbers as they are, which means that we'll consider that an app with 100,000+ installs has 100,000 installs, and an app with 1,000,000+ installs has 1,000,000 installs, and so on.

To perform computations, however, we'll need to convert each install number to int() — this means that we need to remove the commas and the plus characters, otherwise the conversion will fail and raise an error. We'll do this directly in the loop below, where we also compute the average number of installs for each genre (category).

In [27]:
category_android = f_table(android_final, 1)
from operator import itemgetter

android_installs_avg_cat = {}

for genre in category_android:
    tot_installs = 0
    tot_genre = 0
    for app in android_final:
        
        if genre == app[1]:
           
            installs = str(app[5])
            installs = installs.replace(',', '')
            installs = installs.replace('+', '')
            installs_num = int(installs)
            
            tot_installs += installs_num
            tot_genre += 1
            
    android_installs_avg_cat[genre] = tot_installs / tot_genre
    
android_installs_avg_cat_sorted = sorted(android_installs_avg_cat.items(), key=itemgetter(1), reverse=True)
for genre in android_installs_avg_cat_sorted:
    print(f'{genre[0]} = {genre[1]:.2f}')

COMMUNICATION = 38456119.17
VIDEO_PLAYERS = 24727872.45
SOCIAL = 23253652.13
PHOTOGRAPHY = 17840110.40
PRODUCTIVITY = 16787331.34
GAME = 15588015.60
TRAVEL_AND_LOCAL = 13984077.71
ENTERTAINMENT = 11640705.88
TOOLS = 10801391.30
NEWS_AND_MAGAZINES = 9549178.47
BOOKS_AND_REFERENCE = 8767811.89
SHOPPING = 7036877.31
PERSONALIZATION = 5201482.61
WEATHER = 5074486.20
HEALTH_AND_FITNESS = 4188821.99
MAPS_AND_NAVIGATION = 4056941.77
FAMILY = 3695641.82
SPORTS = 3638640.14
ART_AND_DESIGN = 1986335.09
FOOD_AND_DRINK = 1924897.74
EDUCATION = 1833495.15
BUSINESS = 1712290.15
LIFESTYLE = 1437816.27
FINANCE = 1387692.48
HOUSE_AND_HOME = 1331540.56
DATING = 854028.83
COMICS = 817657.27
AUTO_AND_VEHICLES = 647317.82
LIBRARIES_AND_DEMO = 638503.73
PARENTING = 542603.62
BEAUTY = 513151.89
EVENTS = 253542.22
MEDICAL = 120550.62


On average, communication apps have the most installs: 38,456,119. This number is heavily skewed up by a few apps that have over one billion installs (WhatsApp, Facebook Messenger, Skype, Google Chrome, Gmail, and Hangouts), and a few others with over 100 and 500 million installs:

In [28]:
for app in android_final:
    if app[1] == 'COMMUNICATION' and (app[5] == '1,000,000,000+' or app[5] == '500,000,000+' or app[5] == '100,000,000+'):
        print(f'{app[0]} = {app[5]}')

WhatsApp Messenger = 1,000,000,000+
imo beta free calls and text = 100,000,000+
Android Messages = 100,000,000+
Google Duo - High Quality Video Calls = 500,000,000+
Messenger – Text and Video Chat for Free = 1,000,000,000+
imo free video calls and chat = 500,000,000+
Skype - free IM & video calls = 1,000,000,000+
Who = 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji = 100,000,000+
LINE: Free Calls & Messages = 500,000,000+
Google Chrome: Fast & Secure = 1,000,000,000+
Firefox Browser fast & private = 100,000,000+
UC Browser - Fast Download Private & Secure = 500,000,000+
Gmail = 1,000,000,000+
Hangouts = 1,000,000,000+
Messenger Lite: Free Calls & Messages = 100,000,000+
Kik = 100,000,000+
KakaoTalk: Free Calls & Text = 100,000,000+
Opera Mini - fast web browser = 100,000,000+
Opera Browser: Fast and Secure = 100,000,000+
Telegram = 100,000,000+
Truecaller: Caller ID, SMS spam blocking & Dialer = 100,000,000+
UC Browser Mini -Tiny Fast Private & Secure = 100,000,000+
Viber Mess

If we removed all the communication apps that have over 100 million installs, the average would be reduced roughly ten times:

In [29]:
under_100m = []
for app in android_final:
    installs = app[5]
    installs = installs.replace(',', '')
    installs = installs.replace('+', '')
    installs = int(installs)
    if app[1] == 'COMMUNICATION' and installs < 100000000:
        under_100m.append(installs)

avg_under100m = sum(under_100m) / len(under_100m)
print(avg_under100m)

3603485.3884615386


We see the same pattern for the video players category, which is the runner-up with 24,727,872 installs. The market is dominated by apps like Youtube, Google Play Movies & TV, or MX Player. The pattern is repeated for social apps (where we have giants like Facebook, Instagram, Google+, etc.), photography apps (Google Photos and other popular photo editors), or productivity apps (Microsoft Word, Dropbox, Google Calendar, Evernote, etc.).

Again, the main concern is that these app genres might seem more popular than they really are. Moreover, these niches seem to be dominated by a few giants who are hard to compete against.

The game genre seems pretty popular, but previously we found out this part of the market seems a bit saturated, so we'd like to come up with a different app recommendation if possible.

The books and reference genre looks fairly popular as well, with an average number of installs of 8,767,811. It's interesting to explore this in more depth, since we found this genre has some potential to work well on the App Store, and our aim is to recommend an app genre that shows potential for being profitable on both the App Store and Google Play.

Let's take a look at some of the apps from this genre and their number of installs:

In [30]:
for app in android_final:
    if app[1] == 'BOOKS_AND_REFERENCE':
        print(f'{app[0]} = {app[5]}')

E-Book Read - Read Book for free = 50,000+
Download free book with green book = 100,000+
Wikipedia = 10,000,000+
Cool Reader = 10,000,000+
Free Panda Radio Music = 100,000+
Book store = 1,000,000+
FBReader: Favorite Book Reader = 10,000,000+
English Grammar Complete Handbook = 500,000+
Free Books - Spirit Fanfiction and Stories = 1,000,000+
Google Play Books = 1,000,000,000+
AlReader -any text book reader = 5,000,000+
Offline English Dictionary = 100,000+
Offline: English to Tagalog Dictionary = 500,000+
FamilySearch Tree = 1,000,000+
Cloud of Books = 1,000,000+
Recipes of Prophetic Medicine for free = 500,000+
ReadEra – free ebook reader = 1,000,000+
Anonymous caller detection = 10,000+
Ebook Reader = 5,000,000+
Litnet - E-books = 100,000+
Read books online = 5,000,000+
English to Urdu Dictionary = 500,000+
eBoox: book reader fb2 epub zip = 1,000,000+
English Persian Dictionary = 500,000+
Flybook = 500,000+
All Maths Formulas = 1,000,000+
Ancestry = 5,000,000+
HTC Help = 10,000,000+
E

The book and reference genre includes a variety of apps: software for processing and reading ebooks, various collections of libraries, dictionaries, tutorials on programming or languages, etc. It seems there's still a small number of extremely popular apps that skew the average:

In [31]:
for app in android_final:
    installs = app[5]
    installs = installs.replace(',', '')
    installs = installs.replace('+', '')
    installs = int(installs)
    
    if app[1] == 'BOOKS_AND_REFERENCE' and installs >= 100000000:
        print(f'{app[0]} = {app[5]}')

Google Play Books = 1,000,000,000+
Bible = 100,000,000+
Amazon Kindle = 100,000,000+
Wattpad 📖 Free Books = 100,000,000+
Audiobooks from Audible = 100,000,000+


However, it looks like there are only a few very popular apps, so this market still shows potential. Let's try to get some app ideas based on the kind of apps that are somewhere in the middle in terms of popularity (between 1,000,000 and 100,000,000 downloads):

In [32]:
for app in android_final:
    installs = app[5]
    installs = installs.replace(',', '')
    installs = installs.replace('+', '')
    installs = int(installs)
    
    if app[1] == 'BOOKS_AND_REFERENCE' and 1000000 <= installs < 100000000:
        print(f'{app[0]} = {app[5]}')

Wikipedia = 10,000,000+
Cool Reader = 10,000,000+
Book store = 1,000,000+
FBReader: Favorite Book Reader = 10,000,000+
Free Books - Spirit Fanfiction and Stories = 1,000,000+
AlReader -any text book reader = 5,000,000+
FamilySearch Tree = 1,000,000+
Cloud of Books = 1,000,000+
ReadEra – free ebook reader = 1,000,000+
Ebook Reader = 5,000,000+
Read books online = 5,000,000+
eBoox: book reader fb2 epub zip = 1,000,000+
All Maths Formulas = 1,000,000+
Ancestry = 5,000,000+
HTC Help = 10,000,000+
Moon+ Reader = 10,000,000+
English-Myanmar Dictionary = 1,000,000+
Golden Dictionary (EN-AR) = 1,000,000+
All Language Translator Free = 1,000,000+
Aldiko Book Reader = 10,000,000+
Dictionary - WordWeb = 5,000,000+
50000 Free eBooks & Free AudioBooks = 5,000,000+
Al-Quran (Free) = 10,000,000+
Al Quran Indonesia = 10,000,000+
Al'Quran Bahasa Indonesia = 10,000,000+
Al Quran Al karim = 1,000,000+
Al Quran : EAlim - Translations & MP3 Offline = 5,000,000+
Koran Read &MP3 30 Juz Offline = 1,000,000+
H

This niche seems to be dominated by software for processing and reading ebooks, as well as various collections of libraries and dictionaries, so it's probably not a good idea to build similar apps since there'll be some significant competition.

We also notice there are quite a few apps built around the book Quran, which suggests that building an app around a popular book can be profitable. It seems that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets.

However, it looks like the market is already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.

## Conclusions
In this project, we analyzed data about the App Store and Google Play mobile apps with the goal of recommending an app profile that can be profitable for both markets.

We concluded that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets. The markets are already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.